routing/endpointregistry.go

```diff
@@ -55,24 +89,23 @@ func (r *EndpointRegistry) Do(routes []*Route) []*Route {
 			for _, epi := range route.LBEndpoints {
 				metrics := r.GetMetrics(epi.Host)
 				if metrics.DetectedTime().IsZero() {
-					r.SetDetectedTime(epi.Host, now)
+					metrics.SetDetected(now)
 				}
 
-				r.lastSeen[epi.Host] = now
+				metrics.SetLastSeen(now)
 			}
 		}
 	}
 
-	for host, ts := range r.lastSeen {
-		if ts.Add(r.lastSeenTimeout).Before(now) {
-			r.mu.Lock()
-			if r.data[host].inflightRequests == 0 {
-				delete(r.lastSeen, host)
-				delete(r.data, host)
-			}
-			r.mu.Unlock()
+	removeOlder := now.Add(-r.lastSeenTimeout)
+	r.data.Range(func(key, value any) bool {
+		e := value.(*entry)
+		if e.LastSeen().Before(removeOlder) {
+			r.data.Delete(key)
```

Comment:
After this we could `return false` to stop the range early, or do I miss something?

Reply:
Ah no we clean up all entries of the map older than that.